File: video_send_recv/server/video_api.py
"""
Simple Video API - Easy-to-use functions for video communication.

This module provides simple functions that other modules can import and use.
"""

import asyncio
import logging
from typing import Optional, Callable, Dict, Any
from dataclasses import dataclass

from aiortc import RTCSessionDescription
from video_communication import VideoCommunicationManager, IntensityData, VideoConfig

logger = logging.getLogger(__name__)

# Global video manager instance
_video_manager: Optional[VideoCommunicationManager] = None


def init_video_system(config: VideoConfig = None) -> VideoCommunicationManager:
    """Initialize the video communication system."""
    global _video_manager
    _video_manager = VideoCommunicationManager(config)
    logger.info("Video communication system initialized")
    return _video_manager


def get_video_manager() -> VideoCommunicationManager:
    """Get the global video manager instance."""
    if _video_manager is None:
        raise RuntimeError("Video system not initialized. Call init_video_system() first.")
    return _video_manager


async def create_connection(connection_id: str):
    """Create a new video connection."""
    manager = get_video_manager()
    return await manager.init_connection(connection_id)


async def connect_video(connection_id: str, sdp: str, sdp_type: str) -> Dict[str, str]:
    """Handle WebRTC offer and create answer."""
    manager = get_video_manager()
    
    # Create connection
    pc = await create_connection(connection_id)
    
    # Set remote description
    offer = RTCSessionDescription(sdp=sdp, type=sdp_type)
    await pc.setRemoteDescription(offer)
    
    # Create answer
    answer = await pc.createAnswer()
    await pc.setLocalDescription(answer)
    
    return {
        "sdp": pc.localDescription.sdp,
        "type": pc.localDescription.type
    }


def register_intensity_handler(connection_id: str, handler: Callable[[IntensityData], None]):
    """Register a handler for intensity data from a specific connection."""
    manager = get_video_manager()
    manager.register_intensity_callback(connection_id, handler)


def unregister_intensity_handler(connection_id: str):
    """Unregister intensity handler for a connection."""
    manager = get_video_manager()
    manager.unregister_intensity_callback(connection_id)


async def disconnect_video(connection_id: str):
    """Disconnect a video connection."""
    manager = get_video_manager()
    await manager.cleanup_connection(connection_id)


def get_connection_info(connection_id: str) -> Optional[Dict[str, Any]]:
    """Get information about a specific connection."""
    manager = get_video_manager()
    return manager.get_connection_stats(connection_id)


def get_all_connections() -> Dict[str, Dict[str, Any]]:
    """Get information about all active connections."""
    manager = get_video_manager()
    return manager.get_all_connections()


async def shutdown_video_system():
    """Shutdown the entire video system."""
    manager = get_video_manager()
    await manager.cleanup_all_connections()
    logger.info("Video system shutdown complete")


# Convenience functions for common use cases
async def start_video_processing(connection_id: str, 
                                intensity_callback: Callable[[IntensityData], None] = None):
    """Start video processing for a connection with optional callback."""
    if intensity_callback:
        register_intensity_handler(connection_id, intensity_callback)
    logger.info(f"Started video processing for {connection_id}")


def is_connection_active(connection_id: str) -> bool:
    """Check if a connection is active."""
    manager = get_video_manager()
    return connection_id in manager.connections


def get_connection_count() -> int:
    """Get the number of active connections."""
    manager = get_video_manager()
    return len(manager.connections)


# Example usage functions
def example_intensity_handler(data: IntensityData):
    """Example intensity handler that logs the data."""
    logger.info(f"Intensity for {data.connection_id}: "
                f"Current={data.current_intensity:.1f}, "
                f"Average={data.average_intensity:.1f}")


async def example_usage():
    """Example of how to use the video API."""
    # Initialize the system
    init_video_system()
    
    # Create a connection
    connection_id = "example_conn"
    pc = await create_connection(connection_id)
    
    # Register intensity handler
    register_intensity_handler(connection_id, example_intensity_handler)
    
    # Start processing
    await start_video_processing(connection_id, example_intensity_handler)
    
    # Get connection info
    info = get_connection_info(connection_id)
    print(f"Connection info: {info}")
    
    # Cleanup
    await disconnect_video(connection_id)

File: video_send_recv/web-app/src/video-communication/VideoAPI.ts
/**
 * Simple Video API - Easy-to-use functions for video communication.
 * 
 * This module provides simple functions that other modules can import and use.
 */

import { 
  VideoCommunicationManager, 
  VideoConfig, 
  VideoConnection, 
  VideoConnectionEvents 
} from './VideoCommunicationManager';
import { IntensityMetrics } from '../types';

// Global video manager instance
let videoManager: VideoCommunicationManager | null = null;

/**
 * Initialize the video communication system.
 */
export function initVideoSystem(config: VideoConfig): VideoCommunicationManager {
  videoManager = new VideoCommunicationManager(config);
  console.log('Video communication system initialized');
  return videoManager;
}

/**
 * Get the global video manager instance.
 */
export function getVideoManager(): VideoCommunicationManager {
  if (!videoManager) {
    throw new Error('Video system not initialized. Call initVideoSystem() first.');
  }
  return videoManager;
}

/**
 * Create a new video connection.
 */
export async function createConnection(connectionId: string): Promise<VideoConnection> {
  const manager = getVideoManager();
  return await manager.createConnection(connectionId);
}

/**
 * Start camera and get media stream.
 */
export async function startCamera(connectionId: string): Promise<MediaStream> {
  const manager = getVideoManager();
  return await manager.startCamera(connectionId);
}

/**
 * Connect to the video server.
 */
export async function connectVideo(connectionId: string, stream: MediaStream): Promise<void> {
  const manager = getVideoManager();
  return await manager.connect(connectionId, stream);
}

/**
 * Disconnect a video connection.
 */
export async function disconnectVideo(connectionId: string): Promise<void> {
  const manager = getVideoManager();
  return await manager.disconnect(connectionId);
}

/**
 * Set event handlers for video communication.
 */
export function setEventHandlers(events: VideoConnectionEvents): void {
  const manager = getVideoManager();
  manager.setEventHandlers(events);
}

/**
 * Get connection information.
 */
export function getConnectionInfo(connectionId: string): VideoConnection | undefined {
  const manager = getVideoManager();
  return manager.getConnection(connectionId);
}

/**
 * Get all connections.
 */
export function getAllConnections(): VideoConnection[] {
  const manager = getVideoManager();
  return manager.getAllConnections();
}

/**
 * Check if a connection is active.
 */
export function isConnectionActive(connectionId: string): boolean {
  const manager = getVideoManager();
  return manager.isConnectionActive(connectionId);
}

/**
 * Get connection count.
 */
export function getConnectionCount(): number {
  const manager = getVideoManager();
  return manager.getConnectionCount();
}

/**
 * Cleanup all connections.
 */
export async function cleanupVideoSystem(): Promise<void> {
  const manager = getVideoManager();
  return await manager.cleanup();
}

/**
 * Convenience function to start video processing for a connection.
 */
export async function startVideoProcessing(
  connectionId: string,
  intensityCallback?: (connectionId: string, metrics: IntensityMetrics) => void
): Promise<void> {
  const manager = getVideoManager();
  
  if (intensityCallback) {
    manager.setEventHandlers({
      onIntensityUpdate: intensityCallback
    });
  }
  
  console.log(`Started video processing for ${connectionId}`);
}

/**
 * Get default video configuration.
 */
export function getDefaultConfig(): VideoConfig {
  return {
    serverUrl: 'http://localhost:8080',
    videoConstraints: {
      width: { ideal: 640, max: 640 },
      height: { ideal: 360, max: 360 },
      frameRate: { ideal: 30, max: 30 },
      facingMode: 'user'
    },
    dataChannelConfig: {
      ordered: false,
      maxRetransmits: 0,
      protocol: 'intensity-v1'
    },
    reconnectInterval: 5000,
    maxReconnectAttempts: 3
  };
}

/**
 * Example intensity handler that logs the data.
 */
export function exampleIntensityHandler(connectionId: string, data: IntensityMetrics): void {
  console.log(`Intensity for ${connectionId}: Current=${data.intensity.toFixed(1)}, Average=${data.avg_intensity.toFixed(1)}`);
}

/**
 * Example usage function.
 */
export async function exampleUsage(): Promise<void> {
  console.log('=== Video API Usage Example ===');
  
  // Initialize the system
  const config = getDefaultConfig();
  initVideoSystem(config);
  
  // Create a connection
  const connectionId = 'example_conn';
  await createConnection(connectionId);
  
  // Set up event handlers
  setEventHandlers({
    onIntensityUpdate: exampleIntensityHandler,
    onConnectionStateChange: (id, state) => {
      console.log(`Connection ${id} state: ${state}`);
    },
    onError: (id, error) => {
      console.error(`Error for ${id}:`, error);
    }
  });
  
  // Start camera
  const stream = await startCamera(connectionId);
  console.log('Camera started:', stream);
  
  // Connect to server
  await connectVideo(connectionId, stream);
  console.log('Connected to server');
  
  // Get connection info
  const info = getConnectionInfo(connectionId);
  console.log('Connection info:', info);
  
  // Cleanup
  await disconnectVideo(connectionId);
  console.log('Disconnected');
}
